### ISSUE TITLE
New Rule: Consistent Multiline ON clause indentation

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#2538]
This is a suggestion for a potential rule where a join clause spans multiple lines.

## Trivial case

Single line joins are easy:

```sql
inner join foo on (bar.a = foo.a)
```

I don't propose that raises any issues.

## Multiline ON conditions.

The following are all valid (and comply with existing indentation rules):

option 1 (no brackets, `on` is on same line as `join`)
```sql
inner join foo on
    bar.a = foo.a
    and bar.b = foo.b
left join ...
```

option 2 (bracketed, `on` is on same line as `join`). Note the extra line before the next join.
```sql
inner join foo on (
    bar.a = foo.a
    and bar.b = foo.b
)
left join ...
```

option 3 (not bracketed, `on` is on newline and rest of conditions indented):
```sql
inner join foo
    on bar.a = foo.a
        and bar.b = foo.b
left join ...
```

I've noticed inconsistency here within our current projects, and I think it's probably unhelpful as they all look slightly different and read differently. I'm not sure which is _best_, although I'm personally leaning towards option 1, because it feels the simplest and cleanest.

Do people think this is a good rule or is it overkill?
If it's good - should we just enforce option 1 or should it be confiugrable?

**Comment by elyobo** [comment]
I don't know what the policy is on new rules and customisation, but I would use this if it was implemented. Personal preference would be 2 then 1 (much the same, I just find the brackets help me to read them and it's consistent with `USING` which continues to require them) and absolutely under no circumstances never 3.

**Comment by jpy-git** [comment]
Weird, in my experience option 3 is very common and not an anti-pattern (I've personally never seen someone use options 1 or 2)

e.g. Snowflake docs
<img width="320" alt="image" src="https://user-images.githubusercontent.com/80432516/156550739-1a6499aa-c1ae-4a39-966c-921b0cf7513c.png">

So should definitely be configurable if implemented

**Comment by elyobo** [comment]
I think developers can adjust to almost anything, so long as it's consistent ;)

Agreed that it should be configurable but I don't think indenting the from inside the select (as in this snowflake example) is possible at all given current sqlfluff config (and not inside the scope of this issue which is about indentation of on clauses).

**Comment by fmms** [comment]
i would configure option 2 and would be a huge fan of this rule! I was just about to open a ticket, to enforce brackets arround the ON clause.

**Comment by daviewales** [comment]
What about when there is only one `=` clause, but it's too long for the line limit?
By default `sqlfluff` does this:


``` sql
select *
from very_long_table_name_1
inner join very_long_table_name_2
    on
        very_long_table_name_1.long_column_name_1
        = very_long_table_name_2.long_column_name_2
```

But I think I'd want something like an implicit indent, but for operators:

``` sql
select *
from very_long_table_name_1
inner join very_long_table_name_2
    on very_long_table_name_1.long_column_name_1
       = very_long_table_name_2.long_column_name_2
```

**Comment by arasdk** [comment]
Would be a fan of this rule. Joins are for obvious reasons pretty pervasive in any SQL project, and being able to configure the layout would benefit readability and consistency across the project.
